src/realm/sync/trigger.hpp

```diff
@@ -0,0 +1,138 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <realm/status.hpp>
+#include <realm/util/bind_ptr.hpp>
+#include <realm/sync/socket_provider.hpp>
+
+namespace realm::sync {
+
+/// Register a function whose invocation can be triggered repeatedly.
+///
+/// While the function is always executed by the event loop thread, the
+/// triggering of its execution can be done by any thread.
+///
+/// The function is guaranteed to not be called after the Trigger object is
+/// destroyed.
+///
+/// Note that even though the trigger() function is thread-safe, the Trigger
+/// object, as a whole, is not. In particular, construction and destruction must
+/// not be considered thread-safe.
+template <class Service>
+class Trigger final {
+public:
+    Trigger(Service* service, SyncSocketProvider::FunctionHandler&& handler);
+    ~Trigger() noexcept;
+
+    Trigger() noexcept = default;
```

Comment:
It is needed because triggers are class members that are not initialized in the constructor (and also not used though a pointer currently)

Reply:
It looks like they are all initialized in constructors - Unless I'm mistaken, you should be able to move the creation of the trigger to the initializer list. If they are towards the end of the list, any dependencies should already be initialized.
e.g.:
```
    , m_keep_running_timer{get_service()} // Throws
    , m_on_idle{&client.get_service(), [this](Status status) {
        if (!status.is_ok())
            return;
        REALM_ASSERT(m_activated);
        if (m_state == ConnectionState::disconnected && m_num_active_sessions == 0) {
            on_idle(); // Throws
            // Connection object may be destroyed now.
        }
    }};
```

Reply:
I was just double-checking, and it seems that the instance used by the server is complaining although it's also initialized in the constructor..

Reply:
Or, maybe just make sure to check for `m_handler_info` in the trigger function before the second assert.

Reply:
I think we'll get rid off the constructor if we start using the triggers through a pointer, so it will probably happen in the next PR.

I don't understand why the check is needed. The idea is that a trigger should be constructed properly (hence the assertions), the default constructor is there just for convenience.

Reply:
I looked into moving the creation of the triggers to the initializer list. it will require re-ordering the class members so they are initialized in the right order. I would rather not do that now, and as I said, I think we'll remove the default constructor in a later PR.